Google.Api.Generator/Utils/ResourcePattern.cs

```diff
@@ -16,107 +16,154 @@
+                    else if (p.EndsWith("=*"))
                     {
-                        if (inSep)
-                        {
-                            throw new ArgumentException($"Segment '{segment}' is ill-formed; incorrect '}}' found.");
-                        }
-                        var s = sb.ToString();
-                        parameterNamesWithSuffix.Add(s);
-                        if (s.EndsWith("=**"))
-                        {
-                            s = s[0..^3];
-                        }
-                        else if (s.EndsWith("=*"))
-                        {
-                            s = s[0..^2];
-                        }
-                        if (s == "")
-                        {
-                            throw new ArgumentException($"Segment '{segment}' is ill-formed; it is empty.");
-                        }
-                        if (!(s[0] >= 'a' && s[0] <= 'z' && s.All(c => (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_')))
-                        {
-                            throw new ArgumentException($"Segment '{segment}' is ill-formed; it must conform to /[a-z][a-zA-Z0-9_]*/");
-                        }
-                        parameterNames.Add(s);
-                        sb.Clear();
-                        inSep = true;
+                        p = p[..^2];
                     }
-                    else
+                    bool valid = p[0] >= 'a' && p[0] <= 'z' && p.All(c => (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_');
+                    Check(valid, $"parameter name '{p}' contains invalid characters.");
+                    parameterNames.Add(p);
+                    if (closePos == segment.Length - 1)
                     {
-                        sb.Append(c);
+                        break;
                     }
+                    Check(segment.Length >= closePos + 3, "unexpected end of segment.");
+                    char sep = segment[closePos + 1];
+                    Check(sep == '_' || sep == '-' || sep == '.' || sep == '~', $"invalid separator character '{sep}'.");
+                    separators.Add(sep);
+                    pos = closePos + 2;
                 }
-                if (!inSep)
-                {
-                    throw new ArgumentException($"Segment '{segment}' is ill-formed; missing final '}}'.");
-                }
-                separators.Add(sb.ToString());
-                if (separators.Any(x => x != "") && parameterNames.Zip(parameterNamesWithSuffix, (a, b) => (a, b)).Any(x => x.a != x.b))
+                if (parameterNames.Count > 1)
                 {
-                    throw new ArgumentException($"Segment '{segment}' is ill-formed; multi-id segment may not use '=*' or '=**' suffix.");
+                    Check(parameterNamesWithSuffix.All(p => !p.EndsWith("=**")), "'=**' suffix not valid in a multi-ID segment.");
```

Comment:
I don't see how we'd get here anyway - aren't we removing the `=**` in lines 80-83?

Reply:
This is using the `parameterNamesWithSuffix` list, which _does_ contain the suffixes. This check is only relevant if there is more than one parameter-name, hence it's position in this `if` statement.